node/pkg/ethereum/moonbeamimpl.go

```diff
@@ -0,0 +1,259 @@
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+				for {
+					// See if the next block has been created yet.
+					if currentBlockNumber > latestBlockNumber {
+						latestBlockNumber, err = e.getLatestBlockNumber(ctx)
+						if err != nil {
+							e.logger.Error("failed to look up latest block", zap.String("eth_network", e.BaseEth.NetworkName),
+								zap.Uint64("block", currentBlockNumber), zap.Error(err))
+							break
+						}
+
+						if currentBlockNumber > latestBlockNumber {
+							e.logger.Info("waiting for new latest block", zap.String("eth_network", e.BaseEth.NetworkName),
+								zap.Uint64("current_block", currentBlockNumber), zap.Uint64("latest_block", latestBlockNumber))
+							break
+						}
+					}
+
+					// Fetch the hash every time, in case it changes due to a rollback.
+					blockHash, err := e.getBlockHash(ctx, currentBlockNumber)
+					if err != nil {
+						e.logger.Error("failed to look up hash for current block", zap.String("eth_network", e.BaseEth.NetworkName),
+							zap.Uint64("block", currentBlockNumber), zap.Error(err))
+						break
+					}
+
+					e.logger.Info("checking to see if block is finalized", zap.String("eth_network", e.BaseEth.NetworkName),
+						zap.Uint64("block", currentBlockNumber), zap.Stringer("hash", blockHash))
+					finalized, err := e.isBlockFinalized(ctx, blockHash.Hex())
+					if err != nil {
+						e.logger.Error("failed to see if block is finalized", zap.String("eth_network", e.BaseEth.NetworkName),
+							zap.Uint64("block", currentBlockNumber), zap.Error(err))
+						break
+					}
+					if !finalized {
+						break
+					}
+
+					e.logger.Info("block is now finalized", zap.String("eth_network", e.BaseEth.NetworkName),
+						zap.Uint64("block", currentBlockNumber), zap.Stringer("hash", blockHash))
+
+					ev, err := e.getBlock(ctx, currentBlockNumber)
+					if err != nil {
+						e.logger.Error("failed to get finalized block", zap.String("eth_network", e.BaseEth.NetworkName),
+							zap.Uint64("block", currentBlockNumber), zap.Error(err))
+						// Don't break, move on to the next block.
+					} else {
+						sink <- &common.NewBlock{
+							Number: ev.Number,
+							Hash:   blockHash,
+						}
+					}
+
+					currentBlockNumber += 1
+				}
+
+				timer = time.NewTimer(DELAY_IN_MS * time.Millisecond)
+			}
+		}
+	}()
+
+	sub := &MoonbeamSubscription{
+		err: make(chan error, 1),
+	}
+
+	return sub, err
+}
+
+func (e *MoonbeamImpl) getLatestBlockNumber(ctx context.Context) (uint64, error) {
+	type Marshaller struct {
+		Number *ethHexUtils.Big
+	}
+
+	var m Marshaller
+	err := e.mbRpcCon.CallContext(ctx, &m, "eth_getBlockByNumber", "latest", false)
+	if err != nil {
+		e.logger.Error("failed to get current block", zap.String("eth_network", e.BaseEth.NetworkName), zap.Error(err))
+		return 0, err
+	}
+
+	return strconv.ParseUint(m.Number.String()[2:], 16, 64)
+}
+
+func (e *MoonbeamImpl) getBlockHash(ctx context.Context, number uint64) (ethCommon.Hash, error) {
```

Comment:
I think there's a way to combine this with `getLatestBlockNumber`, check out https://github.com/ethereum/go-ethereum/blob/440c9fcf75d9d5383b72646a65d5e21fa7ab6a26/ethclient/ethclient.go#L533-L542. Also should number be `*big.Int`?